utils: init genSecretsReplacement (#503858)
NixOS/nixpkgs: doc/functions/secrets.section.md

as that file's content is stored in the nix store.

Secrets should instead be referenced by path `"/path/to/a/file"` with modules providing options accepting the `string` type.
Those files need to be then copied to or generated on
the target host by out of band tools such as those listed [in the wiki](https://wiki.nixos.org/wiki/Comparison_of_secret_managing_schemes).

## Configuration Files {#sec-secrets-configuration-files}

In the case of services requiring the secrets to live inside their configuration file,
nixpkgs provides the `genSecretsReplacement` function
and its `gen*SecretsReplacement` specializations to help modules implement this safely with minimal effort.
These functions generate configuration files with secrets embedded in clear text
while avoiding storing the clear text secrets in the nix store.
Out of band tools are still needed to copy or generate those secrets on the target host as mentioned above.

The functions expect multiple arguments which are covered in details in their respective documentation.
To follow this section, it is enough to know they accept:

- The configuration file as nix expression,

> - a generator function `a -> string` to generate the configuration from in wanted format,

Reviewer: ```suggestion
- a generator function `a -> string` to serialize the configuration from Nix to the wanted format,
```

- the location where the final configuration with embedded secrets should live,

> - and others for fine tuning.

Reviewer: ```suggestion
- extra options, relevant to specific fomats, to fine-tune their behaviour and output.
```

> The nix expression representing the configuration is an arbitrary value of nested lists and attrsets - which is fully compatible

Reviewer: ```suggestion
The nix expression representing the configuration is an arbitrary value of atoms, nested lists and attrsets - which is fully compatible

with [freeform type options](https://nixos.org/nixos/manual/index.html#sec-freeform-modules) -
where secrets are identified as an attrset containing the `_secret` attr:

```nix
{
  mysecret = {
    _secret = "/run/secrets/path/to/secret";
  };
}
```

The above `{ _secret = "/run/secrets/path/to/secret" }` attrset defines
where the decrypted secret will be available from on the target host.

The function then replaces all occurrences of such an attrset with a uniquely identified placeholder
string whose exact value is an implementation detail:

```nix
{

>   mysecret = "a unique placholder";

Reviewer: Maybe:

```suggestion
  mysecret = /* some unique placeholder */;
```

Author: Actually this does not pass the markdown linter. I'll replace by:

```nix
  mysecret = "<some unique placeholder>";
```

}
```

From this new representation, the function generates the configuration file in the expected output format
and this gets stored in the nix store.
For example in JSON format:

```json
{

>   "mysecret": "a unique placholder"

Reviewer: Maybe:

```suggestion
  "mysecret": <a unique placeholder>
```

Reviewer: Otherwise even maybe just use the actual placeholder that is generated 🤷.

Author: The goal was to obfuscate the format used by the placeholder but I’m not sure it’s useful to hide it.

}
```

The function then returns a bash script whose purpose is to replace all those placeholders
with the secret coming from the path in the `_secret` attr.
Usually this bash script is placed inside the `serviceConfig.ExecStartPre` of the systemd service
needing the configuration file. See the next sections for examples.

> Assuming the file `/run/secrets/path/to/secret` exists on the target host and contains

Reviewer: ```suggestion
Assuming the file `/run/secrets/path/to/secret` is readable on the target host and contains
```

`my super secret`, running the bash script returned by the function will output the following configuration
in the expected location:

```json
{
  "mysecret": "my super secret"
}
```

### JSON configuration file example {#sec-secrets-configuration-files-json}

For JSON configuration files, the `genJqSecretsReplacement` function can be used.
The following snippet only shows parts related to using this function
and does not present best practices related to other parts of module creation and maintenance.

```nix
{
  config,
  lib,
  pkgs,
  ...
}:
let
  cfg = config.services.myservice;

  settingsFormat = pkgs.formats.json { };
  settingsFile = "/run/myservice/myconfiguration.json";

  inherit (lib)
    mkOption
    types
    ;
in
{
  options.services.myservice = {
    settings = mkOption {
      type = types.submodule {
        freeformType = settingsFormat.type;
      };
      default = { };

>     };

Reviewer: Maybe: in principle, I would add a line of documentation mentioning `_secret`, somehow?

Author: I added some copy/pastable documention.

  };

  config = {
    systemd.services.myservice =
      let
        secretsReplacement = utils.genJqSecretsReplacement { } cfg.settings settingsFile;

>       in

Reviewer: I'd move this `let ... in` to the top-level, to match how it would actually be used in a module.

Reviewer: After a `grep`, maybe I'm the only one doing that, you can disregard this comment.

Author: I find your way makes the snippet clearer.

      {
        preStart = secretsReplacement.script;
      };
  };
}
```

Note that the systemd service must have permission to read the files containing the secrets.
Either the owner of the files must match the owner of the systemd service,
or systemd's `LoadCredential` feature can be used as shown in the next section.

We recommend against prefixing the `preStart` script with the bang `!`
because this gives root access to the script and is unnecessarily permissive.

### JSON configuration file LoadCredential example {#sec-secrets-configuration-files-json-loadcredential}

To avoid any permission issue for reading the files containing the secrets,
when using systemd the easiest is to use its `LoadCredential` feature.

The `gen*SecretsReplacement` functions can be configured to instead return an attrset
containing the bash script in the `script` attr and what to give to `LoadCredential`

> in the `credentials` attr.

Reviewer: It retuns an _additional_ `credentials` attribute, but it always returns an attrset, right?

It's `genJqSecretsReplacementSnippet` that returns only the script as a string.

# Secrets {#sec-secrets}

Nix stores every file in clear text in the nix store. There is no
encrypted sub-path. This means one should never reference secrets by
using the path syntax `${./path/to/a/file}` as that file's content is
stored in the nix store.

Secrets should instead be referenced with a path string

> `"/path/to/a/file"` with modules providing options accepting the

Reviewer 2: Sorry, maybe not that related, but I thought the recommended way in this case is to use [externalPath](https://nixos.org/manual/nixos/stable/#sec-option-types-basic). Is the other way around? Should we prefer string over `externalPath` for secrets stored in a path?

Author: It seems like a perfect fit!

> A path that is not contained in the Nix store. Typical use cases are: secrets, password or any other external file.

Author: Done, see https://github.com/NixOS/nixpkgs/compare/200b5d03b848f07fcdd13f9f0acf847e88b1afc5..0e36d72d043fc0261bc2a7b2e712b89de03e4365
